tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+
+            -   Must have same sorted "degree sequence" overall nodes
+            -   Must have same # of source nodes
+            -   When performing a DFS nodes simultaneously in both graphs using
+                2 pointers, one in each MetaGraph, starting from the sources and
+                traversing successors sorted by (op, templateId, neighborCount,
+                predecessorCount, successorCount, nodeName), the MetaGraphs
+                differ:
+                -   If the nodes differ by Op, templateId, type, seriesCount
+                -   If the number of ousorted list of successors (nodes at
+                    outgoing edges) differ
+                -   If one MetaGraph revisits a node while the other MetaGraph
+                    does not
+
+        -   Generate a clusterId as `${signature}[${clusterIndex}]`.
+
+            -   Since 2 MetaNodes might have the same signature, but have
+                different internal structure, they may end up in different
+                clusters, e.g.
+
+                -   MetaNode signature "depth=1 |V|=3 |E|=0 [ops] Add=3"
+                -   MetaNode.templateId `signature + '[' + clusters.length +
+                    ']'`
+
+        -   Treat the ids from the final clusters as 'templateId's.
+
+-   Returns
+
+    ```
+    {
+      root: Metanode;
+      libraryFunctions: {
+        [key: string]: LibraryFunctionData;
+      };
+      templates: {
+        "depth=1 |V|=3 |E|=0 [ops] Add=3[0]": {level: 0, nodes: MetaNode[]}
+      } as {[templateId: string]: string[]};
+      /** List of all device names */
+      devices: string[];
+      /** List of all XLA cluster names */
+      xlaClusters: string[];
+      /** True if at least one tensor in the graph has shape information */
+      hasShapeInfo: boolean;
+      /** The maximum size across all meta edges. Used for scaling thickness. */
+      maxMetaEdgeSize: number;
+      getBridgegraph(nodeName: string): graphlib.Graph;
+      ...
+    }
+    ```
+
+### Rendering the view
+
+TODO
+
+### Node extraction
+
+TODO
+
+## Glossary
+
+Communicating the inner workings of the Graph dashboard requires introducing
+some dashboard-specific concepts used throughout the web frontend code. Much of
+this terminology comes from
+[existing literature](https://www.cs.ubc.ca/labs/imager/tr/2008/Archambault_GrouseFlocks_TVCG/grouseFlocksSub.pdf).
+
+-   **GraphDef**: A 'base graph' structure, loosely tied to TensorFlow's
+    [graph.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/graph.proto).
+-   **NodeDef / RawNode**: A 'base node' structure, loosely tied to TensorFlow's
+    [node_def.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/node_def.proto).
+-   **OpNode**: The result of 'normalizing' a raw `NodeDef`.
+-   **BaseEdge**: An edge between two `OpNodes`.
+-   **SlimGraph**: A lightweight structure containing `OpNode`s and `BaseEdge`s.
```

Comment:
Meta question: do you think we need this data structure or can we go to Hierarchy straight from GraphDef? Does it have actual benefit from the abstraction? 

Also, what do you think about the name? Slim was type of the models for TFv1 and it confuses me :\